old-materials/model_viz.py

```diff
@@ -0,0 +1,608 @@
+                        
+                        # Get our specific previous period dependencies if available
+                        prev_period_deps = getattr(self, 'prev_period_deps', set())
+                        
+                        # Modify formula to show time dependencies with star/prime notation
+                        # based on specific dependencies
+                        for dep_var, dep in prev_period_deps:
+                            if dep_var == var_name and dep in formula:
+                                if f" {dep}" in formula:
+                                    formula = formula.replace(f" {dep}", f" {dep}*")
+                                if f"({dep}" in formula:
+                                    formula = formula.replace(f"({dep}", f"({dep}*")
+                                if f"[{dep}" in formula:
+                                    formula = formula.replace(f"[{dep}", f"[{dep}*")
+                        
+                        formulas[var_name] = f"{var_name} = {formula}"
+                    else:
+                        formulas[var_name] = f"{var_name} = [Function]"
+                except:
+                    formulas[var_name] = f"{var_name} = [Complex Function]"
+        
+        # Add parameter values as formulas
+        for param_name, param_value in self.parameters.items():
+            if param_name in self.variables.get('param_vars', set()):
+                formulas[param_name] = f"{param_name} = {param_value}"
+        
+        # Print formulas for validation
+        print("\nEquation:")
+        for var, formula in sorted(formulas.items()):
+            print(f"  {formula}")
+        
+        # Corrected line: assign the formulas to self.formulas
+        self.formulas = formulas
+    
+    def get_agent_for_variable(self, var_name):
+        """Return the entity to which the variable belongs"""
+        # Handle previous period variables - strip the suffix
+        base_var = var_name.replace("*", "")
+        
+        # Remove _prev suffix used internally
+        if base_var.endswith("_prev"):
+            base_var = base_var[:-5]
+        
+        for agent, vars_list in self.agent_attribution.items():
+            if base_var in vars_list:
+                return agent
+        return "other"
+    
+    def create_graph(self, output_file=None, show=True, color_seed=42):
+        """
+        Create a visual graph of the model using pydot and optionally display
+        
+        Parameters:
+        output_file -- Optional output file path
+        show -- Whether to display the image
+        color_seed -- Seed for color generation (default: 42)
+        
+        Returns:
+        pydot.Dot object representing the model structure graph
+        """
+        if not self.variables or not self.dependencies:
+            self.analyze()
+        
+        # Define different shapes for different variable types
+        shape_map = {
+            "shock_vars": "triangle",
```

Comment:
All of this visual style information (about shapes and colors) should not be baked into the `create_graph` method, but rather should be provided as _arguments_ the create_graph method, so that the user can enter what is needed for the style they want.

But the code for these visual styles can be moved into separate functions that get called for default values in case these arguments are not provided.

Ultimately, that style information should be sourced from a configuration file (like a YAML file), not something hard coded in Python.